### PR TITLE
[quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29416 [quant] Make PerChannelMinMaxObserver scriptable using `torch.jit.ignore`
* **#27950 [quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`**

Summary:
att

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D18360139](https://our.internmc.facebook.com/intern/diff/D18360139)

### CODE REVIEW

**Comment by jerryzh168** [comment]
This is ready for review please take a look

File: test/test_quantization.py [hunk 1]
```diff
@@ -696,6 +698,39 @@ def test_single_liner(self):
             [self.calib_data],
             inplace=False)
         result_eager = model_eager(self.calib_data[0][0])
+        torch._C._jit_pass_quant_fusion(model_script._c._get_module('fc1')._get_method('forward').graph)
+        result_script = model_script._c._get_method('forward')(self.calib_data[0][0])
+        self.assertEqual(result_eager, result_script)
+
+    @_tmp_donotuse_dont_inline_everything
```
Comment (ZolotukhinM) [inline test/test_quantization.py]: The default behavior has recently changed to not inline everything by default - please verify that we can remove this decorator.

File: torch/quantization/observer.py [hunk 2]
```diff
@@ -298,7 +299,10 @@ def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
 
     def forward(self, x_orig):
         r"""Records the running minimum and maximum of ``x``."""
-        x = x_orig.detach()
+        return self._forward(x_orig)
```
Comment (ZolotukhinM) [inline torch/quantization/observer.py]: Why do we need this?

File: torch/csrc/jit/script/module.cpp [hunk 3]
```diff
@@ -244,7 +244,7 @@ Module Module::clone_impl(
     } else {
       r.register_attribute(
           s.name,
-          s.value.type(),
+          type()->getAttribute(s.name),
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/module.cpp]: This is an unrelated change.

File: torch/quantization/observer.py [hunk 4]
```diff
@@ -700,9 +704,11 @@ def _non_linear_param_search(self):
         new_max = self.min_val + bin_width * (end_bin + 1)
         return new_min, new_max
 
+    @torch.jit.ignore
     def _combine_histograms(
         self, dst_histogram, dst_min, dst_max, src_histogram, src_min, src_max
     ):
+        # type: (Tensor, float, float, Tensor, float, float) -> Tensor
```
Comment (ZolotukhinM) [inline torch/quantization/observer.py]: Do you need this if the function is ignored?

File: torch/quantization/observer.py [hunk 5]
```diff
@@ -750,46 +756,50 @@ def _combine_histograms(
             # remaining should go to dst_bin2
             if dst_bin_cnt < src_bin_count:
                 dst_histogram[dst_bin2] += src_bin_count - dst_bin_cnt
+        return dst_histogram
```
Comment (ZolotukhinM) [inline torch/quantization/observer.py]: This seems to be an unnecessary change.

File: test/test_quantization.py [hunk 6]
```diff
@@ -696,6 +698,39 @@ def test_single_liner(self):
             [self.calib_data],
             inplace=False)
         result_eager = model_eager(self.calib_data[0][0])
+        torch._C._jit_pass_quant_fusion(model_script._c._get_module('fc1')._get_method('forward').graph)
```
Comment (ZolotukhinM) [inline test/test_quantization.py]: Where does this come from?

**Review by ZolotukhinM — APPROVED** [review]
Looks good, but please remove unnecessary changes and split out unrelated changes and avoid that in future.

File: test/test_quantization.py (hunk 6 above)
Comment (jerryzh168) [reply test/test_quantization.py]: ...ntot sure, will remove

File: torch/quantization/observer.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/observer.py]: oh, we probably don't need this, I was trying to ignore _forward previously, but it turns out that it doesn't work.

File: torch/quantization/observer.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/observer.py]: yeah, we need this otherwise everything will be treated as Tensor

File: torch/quantization/observer.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/observer.py]: this is needed for compile as well, ignored functions seems to take Tensor as return only

File: test/test_quantization.py (hunk 1 above)
Comment (jerryzh168) [reply test/test_quantization.py]: seems to be working
